README.md

[![Arduino CI](https://github.com/RobTillaart/MTP40C/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![JSON check](https://github.com/RobTillaart/MTP40C/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/MTP40C/actions/workflows/jsoncheck.yml)
[![Arduino-lint](https://github.com/RobTillaart/MTP40C/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/MTP40C/actions/workflows/arduino-lint.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/MTP40C/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/MTP40C.svg?maxAge=3600)](https://github.com/RobTillaart/MTP40C/releases)


# MTP40C / MTP40D

Arduino library for MTP40C and MTP40D CO2 sensor.

(include image)


## Description

The library for the MTP40C / MTP40D CO2 sensor is experimental as not all functionality is tested.

Both the MTP40C and MTP40D sensor is an NDIR (Non Dispersive InfraRed) CO2 sensor.

The sensor communicates over a 19200 baud serial (TTL) interface with a microprocessor or PC. 
This implies that calls which can take up to 25 bytes can take as much as about 20 milliseconds.

On the other hand this low baud rate implies it will work over relative long distances.
This signal quality over longer distances is not investigated. 

The MTP40D has more interface options, I2C, PWM and ALARM. 
This library does not support these other interfaces for now.
However minimal examples are added to have a starter but these 
need to be tested if and how well these work.


### Hardware interface


#### MTP40-C

```
               TOPVIEW MTP40-C
              +-------------+---+
              |             | O |
    Vin   1 --|             +---+
    GND   2 --|             |
    TX    3 --|             |
    RX    4 --|             |
    NC    5 --|             +---+
              |             | O |
              +-------------+---+
```

| Pin  | Name  | Description         |
|:----:|:------|:--------------------|
|  1   | Vin   | 4.2V--5.5V          |
|  2   | GND   | idem                |
|  3   | TX    | Transmit 19200 baud |
|  4   | RX    | Receive 19200 baud  |
|  5   | NC    | Not Connected       |


#### MTP40-D

```
               TOPVIEW MTP40-D
              +-------------+
              |             | 
    VCC   5 --|             |-- 1 Vin
    TX    6 --|             |-- 2 GND
    RX    7 --|             |-- 3 ALARM
    NC    8 --|             |-- 4 PWM / I2C
    GND   9 --|             |
              |             |
              +-------------+
```

| Pin  | Name    | Description                 |
|:----:|:--------|:----------------------------|
|  1   | Vin     | 4.2V--5.5V                  |
|  2   | GND     | idem                        |
|  3   | ALARM   | HIGH above 2000 PPM, LOW below 1800 PPM (hysteresis) |
|  4   | PWM/I2C | PWM out or I2C select       |
|  5   | VCC_O   | 3V3 out for serial          |
|  6   | TX      | Transmit 19200 baud or SDA  |
|  7   | RX      | Receive 19200 baud or SCL   |
|  8   | NC      | Not Connected               |
|  9   | GND     | idem                        |


## Interface

### Warnings

During tests with an UNO the communication over Software Serial did fail sometimes.
Therefore it is important to **always check return values** to make your project more robust.

During tests it became clear that the sensor needs time to process 
commands e.g. **setSelfCalibration()**. By having a delay(100) between the calls
everything ran far more stable (within my test). Todo seek optimum delay(), added in Future section below.

The CRC of the sensor responses are not verified by the library.


### Constructors

- **MTP40(Stream \* str)** constructor. should get a Serial port as parameter e.g. \&Serial, \&Serial1. This is the base class.
- **MTP40C(Stream \* str)** constructor. should get a Serial port as parameter e.g. \&Serial, \&Serial1 
or a software Serial port. That Serial port must connect to the sensor. 
- **MTP40D(Stream \* str)** constructor. should get a Serial port as parameter e.g. \&Serial, \&Serial1 
or a software Serial port. That Serial port must connect to the sensor. 
- **bool begin(uint8_t address = 0x64)** initialize the device.
Sets the address to communicate to the sensor. Address values allowed 0 .. 247.
Uses the factory default value of 0x64 when no parameter is given.
Also resets internal settings.
- **bool isConnected()** returns true if the address as set by **begin()** 
or the default address of 0x64 (decimal 100) can be found on the Serial 'bus'.
- **uint8_t getType()** returns 2 for the MTP40C and 3 for the MTP40D sensor.
Return 255 for the MTP40 base class.


### CO2 Measurement

- **uint16_t getGasConcentration()** returns the CO2 concentration in PPM (parts per million).
The function returns **MTP40_INVALID_GAS_LEVEL** if the request fails.

- **void suppressError(bool se)** sets or clears a flag that replaces the error value with 
the last read value if the request fails.
This is useful when plotting the values and one do not want a sudden spike.
One can still check **lastError()** to see if the value was OK.
- **bool getSuppressError()** gets the value of the suppress flag. 
- **int lastError()** returns last error set by **getGasConcentration()** 
or by **getAirPressureReference()** 
Reading resets internal error to MTP40_OK;


### Configuration

- **uint8_t getAddress()** request the address from the device.
Expect a value from 0 .. 247.
Returns **MTP40_INVALID_ADDRESS** (0xFF) if the request fails.
- **bool setAddress(uint8_t address = 0x64)** set a new address for the device. 
0x64 as default. Returns false if not successful. 
If **setSpecificAddress()** is called, this specific address will be used for further commands.

These address functions are only needed if handling multiple devices. (to be tested)
- **void setGenericAddress()** uses the broadcast address 0xFE in all requests. 
This is the default behaviour of the library.
- **void setSpecificAddress()** uses the address specified in **begin()** or 
**setAddress()** or the default 0x64 in all requests.
- **bool useSpecificAddress()** returns true if the specific address is used.
Returns false if the generic / broadcast address is used.

The library can set a maximum timeout in the communication with the sensor.
Normally this is not needed to set as the default of 100 milliseconds is long enough
for even the longest command. This timeout is needed if the sensor did not read the 
command correctly, preventing the host to wait indefinitely.
- **void setTimeout(uint32_t to = 100)** sets the timeout. 
If no parameter is given a default timeout of 100 milliseconds is set.
- **uint32_t getTimeout()** get the value set above or the default. 
Value returned is time in milliseconds.


### Calibration

Please read datasheet before using these functions to understand the process of calibration.


#### Air pressure calibration

- **float getAirPressureReference()** returns the air pressure reference from the device.
Returns **MTP40_INVALID_AIR_PRESSURE** in case request fails.
Default is 1013.0.
- **bool setAirPressureReference(float apr)** to calibrate the air pressure.
One can calibrate the sensor with an external device.
Value for air pressure should between 700.0 and 1100.0. 
The function returns **false** if the parameter is out of range or if the request fails.


#### SPC calibration

This takes a relative short time (few minutes) to calibrate the sensor in a known 
gas concentration. 

- **bool setSinglePointCorrection(float spc)** takes several minutes. see datasheet.
spc should be between 400 and 5000.
The function returns **false** if the parameter is out of range or if the request fails.
- **bool getSinglePointCorrectionReady()** To see if SPC is finished or not. The call also fails if the request fails.

As far as known the SPC point can not be retrieved from the sensor.


#### Self calibration

Self calibration is a process in which the sensor takes the minimum values over a longer period
between 24 - 720 hours as the reference for minimum outdoor values.
Note that 720 hours is 30 days / 1 month.

- **bool openSelfCalibration()** start the self calibration cycle.
- **bool closeSelfCalibration()** stop the self calibration cycle.
- **uint8_t getSelfCalibrationStatus()** Returns if the selfCalibration is open or closed.
**WARNING**: in our test the values in the datasheet seems to be not in sync with the sensor used. 
The function returned **0x00 for CLOSED and 0xFF for OPEN**.
- **bool setSelfCalibrationHours(uint16_t hrs)** Sets the number of hours between self calibration
moments. Valid values are 24 - 720 .
- **uint16_t getSelfCalibrationHours()** returns the value set above.


## Future

#### CRC

- CRC in PROGMEM
- CRC test responses sensor

#### Performance

- performance measurements
- optimize performance if possible
- caching? what?
- seek optimum delay() between calls.

#### Other

- serial bus with multiple devices? => diodes


## Operations

See examples.


## Sponsor 

The development of this MTP40C library is sponsored by [TinyTronics, Netherlands](https://www.tinytronics.nl/shop/nl).


CHANGELOG.md
# Change Log MTP40C

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.3.0] - 2023-07-30
- replace CRC with other algorithm for AVR
  - less RAM, less code, faster (AVR only)
- add MTP40C_CRC_test.ino

----

## [0.2.2] - 2022-11-17
- add RP2040 in build-CI
- add changelog.md

## [0.2.1] - 2021-12-22
- update library.json
- update license
- minor edits


## [0.2.0] - 2021-08-27
- add MTP40D derived class
- many fixes after testing

----

## [0.1.1] - 2021-08-23
- addexamples
- minor fixes

## [0.1.0] - 2021-08-20
- initial version.



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.0
//    DATE: 2021-08-20
// PURPOSE: unit tests for the MTP40C CO2 sensor
//          https://github.com/RobTillaart/MTP40C
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// https://github.com/Arduino-CI/arduino_ci/blob/master/cpp/unittest/Assertion.h#L33-L42
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)



////////////////////////////////////
//
// MANY TESTS WILL BLOCK AS BUILD CI HAS NO GOOD TIMEOUT
// ALL FAILING TESTS ARE COMMENTED
//
// USE GODMODE SERIAL TO IMPROVE THESE TESTS
//

#include <ArduinoUnitTests.h>

#include "Arduino.h"
#include "MTP40C.h"
#include "SoftwareSerial.h"


unittest_setup()
{
}


unittest_teardown()
{
}


unittest(test_getType)
{
  fprintf(stderr, "MTP40_LIB_VERSION:\t%s\n", MTP40_LIB_VERSION);

  MTP40C sensorC = MTP40C(&Serial);
  assertEqual(2, sensorC.getType());

  MTP40D sensorD = MTP40C(&Serial);
  assertEqual(3, sensorD.getType());

  MTP40 sensor = MTP40C(&Serial);
  assertEqual(255, sensor.getType());

}


unittest(test_begin)
{
  fprintf(stderr, "MTP40_LIB_VERSION:\t%s\n", MTP40_LIB_VERSION);

  MTP40C sensor = MTP40C(&Serial);

  for (int addr = 248; addr < 256; addr++)
  {
    assertFalse(sensor.begin(addr));
  }
  // assertTrue(sensor.begin());  blocks!

  // assertFalse(sensor.isConnected());  blocks!
}


unittest(test_address)
{
  fprintf(stderr, "MTP40_LIB_VERSION:\t%s\n", MTP40_LIB_VERSION);

  MTP40C sensor = MTP40C(&Serial);
  // assertTrue(sensor.begin());  blocks!

  // assertEqual(0x64, sensor.getAddress());
  for (int addr = 248; addr < 256; addr++)
  {
    assertFalse(sensor.setAddress(addr));
  }
  // assertFalse(sensor.setAddress(50));
}


unittest(test_air_pressure)
{
  fprintf(stderr, "MTP40_LIB_VERSION:\t%s\n", MTP40_LIB_VERSION);

  MTP40C sensor = MTP40C(&Serial);
  // assertTrue(sensor.begin());    // default address

  // assertEqual(-999, sensor.getAirPressureReference());

  assertFalse(sensor.setAirPressureReference(600.0));
  assertFalse(sensor.setAirPressureReference(1200.0));
  // assertFalse(sensor.setAirPressureReference(1000.0));
}


unittest(test_gas_concentration)
{
  fprintf(stderr, "MTP40_LIB_VERSION:\t%s\n", MTP40_LIB_VERSION);

  MTP40C sensor = MTP40C(&Serial);
  // assertTrue(sensor.begin());

  // assertEqual(0, sensor.getGasConcentration());
}


unittest(test_single_point_correction)
{
  fprintf(stderr, "MTP40_LIB_VERSION:\t%s\n", MTP40_LIB_VERSION);

  MTP40C sensor = MTP40C(&Serial);
  // assertTrue(sensor.begin());

  // assertFalse(sensor.getSinglePointCorrectionReady());

  assertFalse(sensor.setSinglePointCorrection(399.9));
  assertFalse(sensor.setSinglePointCorrection(5000.1));
  // assertFalse(sensor.setSinglePointCorrection(1000.0));
}


unittest(test_self_calibration)
{
  fprintf(stderr, "MTP40_LIB_VERSION:\t%s\n", MTP40_LIB_VERSION);

  MTP40C sensor = MTP40C(&Serial);
  // assertTrue(sensor.begin());    // default address

  // assertFalse(sensor.openSelfCalibration());
  // assertFalse(sensor.closeSelfCalibration());
  // assertEqual(0x02, sensor.getSelfCalibrationStatus());
}


unittest(test_self_calibration_hours)
{
  fprintf(stderr, "MTP40_LIB_VERSION:\t%s\n", MTP40_LIB_VERSION);

  MTP40C sensor = MTP40C(&Serial);
  // assertTrue(sensor.begin());    // default address

  assertFalse(sensor.setSelfCalibrationHours(23));
  assertFalse(sensor.setSelfCalibrationHours(721));
  // assertFalse(sensor.setSelfCalibrationHours(100));
  // assertEqual(0xFFFF, sensor.getSelfCalibrationHours());
}


unittest_main()

// --------


MTP40C.h
#pragma once
//
//    FILE: MTP40C.h
//  AUTHOR: Rob Tillaart
//    DATE: 2021-08-20
// VERSION: 0.3.0
// PURPOSE: Arduino library for MTP40C + MTP40D CO2 sensor
//     URL: https://github.com/RobTillaart/MTP40C
//
// HISTORY: see changelog.md
//
// Based upon datasheet June 2020, version 2.0
//


#include "Arduino.h"


#define MTP40_LIB_VERSION                     (F("0.3.0"))


#define MTP40_DEFAULT_ADDRESS                 0x64

#define MTP40_OK                              0x00
#define MTP40_INVALID_AIR_PRESSURE            0x01
#define MTP40_INVALID_GAS_LEVEL               0x02
#define MTP40_INVALID_ADDRESS                 0xFF


class MTP40
{
public:
  MTP40(Stream * str);

  bool     begin(uint8_t address = MTP40_DEFAULT_ADDRESS);
  bool     isConnected();

  uint8_t  getAddress();
  bool     setAddress(uint8_t address = MTP40_DEFAULT_ADDRESS);

  float    getAirPressureReference();
  bool     setAirPressureReference(float apr);

  uint16_t getGasConcentration();  // returns PPM

  void     suppressError(bool se) { _suppressError = se; };
  bool     getSuppressError()     { return _suppressError; };

  //  CALIBRATION FUNCTIONS
  //  READ DATASHEET !!
  bool     setSinglePointCorrection(float spc);
  bool     getSinglePointCorrectionReady();

  bool     openSelfCalibration();
  bool     closeSelfCalibration();
  uint8_t  getSelfCalibrationStatus();
  bool     setSelfCalibrationHours(uint16_t hours);
  uint16_t getSelfCalibrationHours();

  void     setGenericAddress()  { _useAddress = false; };
  void     setSpecificAddress() { _useAddress = true; };
  bool     useSpecificAddress() { return _useAddress; };

  //  set timeout of serial communication.
  void     setTimeout(uint32_t timeOut = 100) { _timeout = timeOut; };
  uint32_t getTimeout() { return _timeout; };

  uint32_t lastRead() { return _lastRead; };

  //       2   = MTP40C
  //       3   = MTP40D
  //       255 = MTP40 base class
  uint8_t  getType()  { return _type; };
  int      lastError();


//////////////////////////////////////////////////////////////////////
//
//  PROTECTED
//
protected:
  Stream * _ser;
  uint8_t  _buffer[24];   //  should be big enough.
  uint8_t  _address       = MTP40_DEFAULT_ADDRESS;

  bool     _useAddress    = false;
  uint32_t _timeout       = 100;
  uint32_t _lastRead      = 0;

  float    _airPressureReference = 0;
  uint16_t _gasLevel      = 0;
  uint8_t  _type          = 0xFF;

  bool     _suppressError = false;
  int      _lastError     = MTP40_OK;

  bool     request(uint8_t *data, uint8_t cmdlen, uint8_t anslen);

  uint16_t CRC(uint8_t *data, uint16_t len);
  uint8_t  reverse8(uint8_t in);
  uint16_t reverse16(uint16_t in);
};

/////////////////////////////////////////////////////////////
//
// DERIVED CLASSES
//
class MTP40C : public MTP40
{
public:
  MTP40C(Stream * str);
};


class MTP40D : public MTP40
{
public:
  MTP40D(Stream * str);

  // TODO
  // I2C interface
  // PWM interface
};


// -- END OF FILE --



MTP40C.cpp
//
//    FILE: MTP40C.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2021-08-20
// VERSION: 0.3.0
// PURPOSE: Arduino library for MTP40C MTP40D CO2 sensor
//     URL: https://github.com/RobTillaart/MTP40C



#include "MTP40C.h"

//  debug flag, development.
//  #define MTP40_DEBUG    1



MTP40::MTP40(Stream * str)
{
  _ser = str;
  _buffer[0] = '\0';
  _type = 0xFF;
}


bool MTP40::begin(uint8_t address)
{
  if (address > 247) return false;

  _useAddress    = false;
  _timeout       = 100;
  _lastRead      = 0;
  _airPressureReference = 0;
  _gasLevel      = 0;
  _address       = address;
  _suppressError = false;

  return isConnected();
}


bool MTP40::isConnected()
{
  uint8_t addr = getAddress();
  return (addr == _address);
}


uint8_t MTP40::getAddress()
{
  uint8_t cmd[8] = { 0xFE, 0x03, 0x14, 0x00, 0x01, 0x00, 0x55, 0xA5 };
  if (request(cmd, 8, 7) )
  {
    _address = _buffer[3];
    return _buffer[3];
  }

  return MTP40_INVALID_ADDRESS;
}


bool MTP40::setAddress(uint8_t address)
{
  if (address > 247) return false;

  uint8_t cmd[11] = { 0xFE, 0x10, 0x04, 0x00, 0x01, 0x00, 0x01, 0x60, 0x00, 0x42, 0x58 };
  cmd[0] = _address;
  cmd[7] = address;
  if (request(cmd, 11, 8) )
  {
    _address = address;
    return true;
  }
  return false;
}


float MTP40::getAirPressureReference()
{
  union
  {
    float value;
    uint8_t b[4];
  } convert;

  _lastError = MTP40_OK;

  //  max read freq 1x per 4 seconds
  if (millis() - _lastRead < 4000) return _airPressureReference;  // last value
  _lastRead = millis();

  uint8_t cmd[5] = { 0xFE, 0x68, 0x01, 0xFE, 0x30 };
  if (request(cmd, 5, 10))
  {
    for (uint8_t i = 0; i < 4; i++)
    {
      convert.b[i] = _buffer[4 + i];
    }
    _airPressureReference = convert.value;
    return _airPressureReference;
  }

  _lastError = MTP40_INVALID_AIR_PRESSURE;
  if (_suppressError) return _airPressureReference;
  return _lastError;
}


bool MTP40::setAirPressureReference(float apr)
{
  if ((apr < 700) || (apr > 1100)) return false;

  union
  {
    float value;
    uint8_t b[4];
  } convert;

  uint8_t cmd[10] = { 0xFE, 0x67, 0x01, 0x01, 0x00, 0x40, 0x7D, 0x44, 0xC4, 0xA3 };
  convert.value = apr;
  for (uint8_t i = 0; i < 4; i++)
  {
    cmd[4 + i] = convert.b[i];
  }
  if (request(cmd, 10, 10) )
  {
    return true;
  }
  return false;
}


uint16_t MTP40::getGasConcentration()
{
  _lastError = MTP40_OK;

  // max read freq 1x per 4 seconds
  if (millis() - _lastRead < 4000) return _gasLevel;  // last value
  _lastRead = millis();

  uint8_t cmd[5] = { 0xFE, 0x69, 0x03, 0x7E, 0x61 };
  if (request(cmd, 5, 14) )
  {
    // check valid
    for (uint8_t i = 8; i < 12; i++)
    {
      if (_buffer[i] != 0) return 0;
    }
    _gasLevel = _buffer[5] *256 + _buffer[4];
    return _gasLevel;
  }

  _lastError = MTP40_INVALID_GAS_LEVEL;
  if (_suppressError) return _gasLevel;
  return _lastError;
}


bool MTP40::setSinglePointCorrection(float spc)
{
  if ((spc < 400) || (spc > 5000)) return false;

  union
  {
    float value;
    uint8_t b[4];
  } convert;

  uint8_t cmd[9] = { 0xFE, 0x28, 0x80, 0x00, 0x80, 0x40, 0x44, 0x33, 0x22 };
  convert.value = spc;
  for (uint8_t i = 0; i < 4; i++)
  {
    cmd[3 + i] = convert.b[i];
  }
  if (request(cmd, 9, 10) )
  {
    if (_buffer[7] ) return true;
  }
  return false;
}


bool MTP40::getSinglePointCorrectionReady()
{
  uint8_t cmd[5] = { 0xFE, 0x28, 0x81, 0xCE, 0x50 };
  if (request(cmd, 5, 6) )
  {
    if (_buffer[3] == 0) return true;
  }
  return false;
}


bool MTP40::openSelfCalibration()
{
  uint8_t cmd[6] = { 0xFE, 0x28, 0x66, 0xFF, 0xDA, 0x24 };
  if (request(cmd, 6, 6) )
  {
    return true;
  }
  return false;
}


bool MTP40::closeSelfCalibration()
{
  uint8_t cmd[6] = { 0xFE, 0x28, 0x66, 0x00, 0x9A, 0x64 };
  if (request(cmd, 6, 6) )
  {
    return true;
  }
  return false;
}


uint8_t MTP40::getSelfCalibrationStatus()
{
  uint8_t cmd[5] = { 0xFE, 0x28, 0x67, 0x4F, 0xDA };
  if (request(cmd, 5, 6) )
  {
    return _buffer[3];
  }
  return 0x02;
}


bool MTP40::setSelfCalibrationHours(uint16_t hours)
{
  if ((hours < 24) || (hours > 720)) return false;
  uint8_t cmd[7] = { 0xFE, 0x28, 0x6A, 0x64, 0x00, 0x0E, 0xA8 };
  cmd[4] = hours & 0xFF;
  cmd[5] = hours / 256;
  if (request(cmd, 7, 6) )
  {
    return (_buffer[3] == 0x00);
  }
  return false;
}


uint16_t MTP40::getSelfCalibrationHours()
{
  uint8_t cmd[5] = { 0xFE, 0x28, 0x69, 0xCE, 0x1E };
  if (request(cmd, 5, 9) )
  {
    return _buffer[4] * 256 + _buffer[3];
  }
  return 0xFFFF;  // to indicate error.
}


int MTP40::lastError()
{
  int e = _lastError;
  _lastError = MTP40_OK;
  return e;
}



//////////////////////////////////////////////////////////////////////
//
//  PRIVATE
//
bool MTP40::request(uint8_t *data, uint8_t commandLength, uint8_t answerLength)
{
  //  generic or specific address
  if (_useAddress)
  {
    data[0] = _address;
  }
  else
  {
    data[0] = 0xFE;     //  broadcast
  }
  //  calculate CRC of command
  uint16_t crc = CRC(data, commandLength - 2);
  data[commandLength - 1] = crc / 256;
  data[commandLength - 2] = crc & 0xFF;
  while (commandLength--)
  {
#ifdef MTP40_DEBUG
    if (*data < 0x10) _ser->print(0);
    _ser->print(*data++, HEX);
    _ser->print(" ");
#else
    _ser->write(*data++);
#endif
     yield();   //  because baud rate is low!
  }

  uint32_t start = millis();
  uint8_t i = 0;
  while (answerLength)
  {
    if (millis() - start > _timeout) return false;
    if (_ser->available())
    {
      _buffer[i] = _ser->read();
      i++;
      answerLength--;
    }
    yield();  //  because baud rate is low!
  }
  return true;
}


/////////////////////////////////////////////////////////////
//
//  CRC
//
/*
//  from datasheet
uint16_t MTP40::CRC(uint8_t *data, uint16_t len)
{
  //  auchCRCHi contains 2 repeating patterns
  //  0x00, 0xC1, 0x81, 0x40
  //  0x01, 0xC0, 0x80, 0x41
  //  check CRC lib MODBUS polynome (slower and smaller).
const uint8_t auchCRCHi[] = {
  0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81,
  0x40, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0,
  0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01,
  0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41,
  0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81,
  0x40, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0,
  0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01,
  0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40,
  0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81,
  0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0,
  0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01,
  0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
  0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81,
  0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0,
  0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01,
  0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81, 0x40, 0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41,
  0x00, 0xC1, 0x81, 0x40, 0x01, 0xC0, 0x80, 0x41, 0x01, 0xC0, 0x80, 0x41, 0x00, 0xC1, 0x81,
  0x40
};


const uint8_t auchCRCLo[] = {
  0x00, 0xC0, 0xC1, 0x01, 0xC3, 0x03, 0x02, 0xC2, 0xC6, 0x06, 0x07, 0xC7, 0x05, 0xC5, 0xC4,
  0x04, 0xCC, 0x0C, 0x0D, 0xCD, 0x0F, 0xCF, 0xCE, 0x0E, 0x0A, 0xCA, 0xCB, 0x0B, 0xC9, 0x09,
  0x08, 0xC8, 0xD8, 0x18, 0x19, 0xD9, 0x1B, 0xDB, 0xDA, 0x1A, 0x1E, 0xDE, 0xDF, 0x1F, 0xDD,
  0x1D, 0x1C, 0xDC, 0x14, 0xD4, 0xD5, 0x15, 0xD7, 0x17, 0x16, 0xD6, 0xD2, 0x12, 0x13, 0xD3,
  0x11, 0xD1, 0xD0, 0x10, 0xF0, 0x30, 0x31, 0xF1, 0x33, 0xF3, 0xF2, 0x32, 0x36, 0xF6, 0xF7,
  0x37, 0xF5, 0x35, 0x34, 0xF4, 0x3C, 0xFC, 0xFD, 0x3D, 0xFF, 0x3F, 0x3E, 0xFE, 0xFA, 0x3A,
  0x3B, 0xFB, 0x39, 0xF9, 0xF8, 0x38, 0x28, 0xE8, 0xE9, 0x29, 0xEB, 0x2B, 0x2A, 0xEA, 0xEE,
  0x2E, 0x2F, 0xEF, 0x2D, 0xED, 0xEC, 0x2C, 0xE4, 0x24, 0x25, 0xE5, 0x27, 0xE7, 0xE6, 0x26,
  0x22, 0xE2, 0xE3, 0x23, 0xE1, 0x21, 0x20, 0xE0, 0xA0, 0x60, 0x61, 0xA1, 0x63, 0xA3, 0xA2,
  0x62, 0x66, 0xA6, 0xA7, 0x67, 0xA5, 0x65, 0x64, 0xA4, 0x6C, 0xAC, 0xAD, 0x6D, 0xAF, 0x6F,
  0x6E, 0xAE, 0xAA, 0x6A, 0x6B, 0xAB, 0x69, 0xA9, 0xA8, 0x68, 0x78, 0xB8, 0xB9, 0x79, 0xBB,
  0x7B, 0x7A, 0xBA, 0xBE, 0x7E, 0x7F, 0xBF, 0x7D, 0xBD, 0xBC, 0x7C, 0xB4, 0x74, 0x75, 0xB5,
  0x77, 0xB7, 0xB6, 0x76, 0x72, 0xB2, 0xB3, 0x73, 0xB1, 0x71, 0x70, 0xB0, 0x50, 0x90, 0x91,
  0x51, 0x93, 0x53, 0x52, 0x92, 0x96, 0x56, 0x57, 0x97, 0x55, 0x95, 0x94, 0x54, 0x9C, 0x5C,
  0x5D, 0x9D, 0x5F, 0x9F, 0x9E, 0x5E, 0x5A, 0x9A, 0x9B, 0x5B, 0x99, 0x59, 0x58, 0x98, 0x88,
  0x48, 0x49, 0x89, 0x4B, 0x8B, 0x8A, 0x4A, 0x4E, 0x8E, 0x8F, 0x4F, 0x8D, 0x4D, 0x4C, 0x8C,
  0x44, 0x84, 0x85, 0x45, 0x87, 0x47, 0x46, 0x86, 0x82, 0x42, 0x43, 0x83, 0x41, 0x81, 0x80,
  0x40
};

  uint8_t  uchCRCHi = 0xFF ; //  High byte initialization of the CRC
  uint8_t  uchCRCLo = 0xFF ; //  The low byte of the CRC is initialized
  uint16_t uIndex;           //  Query the CRC table index
  uint16_t crc;

  while (len--)               //  Complete the entire message buffer
  {
    uIndex   = uchCRCLo ^ *data++;            //  Calculate CRC
    uchCRCLo = uchCRCHi ^ auchCRCHi[uIndex];
    uchCRCHi = auchCRCLo[uIndex];
  }
  crc = (uint16_t)uchCRCHi * 256;
  crc += (uint16_t)uchCRCLo;
  return crc;
}
*/

//  derived from https://github.com/RobTillaart/CRC 0.3.3
uint16_t MTP40::CRC(uint8_t *array, uint16_t length)
{
  //    parameters           MODBUS
  const uint16_t polynome  = 0x8005;
  const uint16_t startmask = 0xFFFF;
  const uint16_t endmask   = 0x0000;
  const bool reverseIn     = true;
  const bool reverseOut    = true;

  //  start
  uint16_t crc = startmask;
  while (length--)
  {
    if ((length & 0xFF) == 0) yield();  // RTOS

    uint8_t data = *array++;
    if (reverseIn) data = reverse8(data);
    crc ^= ((uint16_t)data) << 8;
    for (uint8_t i = 8; i; i--)
    {
      if (crc & (1 << 15))
      {
        crc <<= 1;
        crc ^= polynome;
      }
      else
      {
        crc <<= 1;
      }
    }
  }
  if (reverseOut) crc = reverse16(crc);
  crc ^= endmask;
  return crc;
}

uint8_t MTP40::reverse8(uint8_t in)
{
  uint8_t x = in;
  x = (((x & 0xAA) >> 1) | ((x & 0x55) << 1));
  x = (((x & 0xCC) >> 2) | ((x & 0x33) << 2));
  x =          ((x >> 4) | (x << 4));
  return x;
}

uint16_t MTP40::reverse16(uint16_t in)
{
  uint16_t x = in;
  x = (((x & 0XAAAA) >> 1) | ((x & 0X5555) << 1));
  x = (((x & 0xCCCC) >> 2) | ((x & 0X3333) << 2));
  x = (((x & 0xF0F0) >> 4) | ((x & 0X0F0F) << 4));
  x = (( x >> 8) | (x << 8));
  return x;
}


/////////////////////////////////////////////////////////////
//
//  DERIVED CLASSES
//

MTP40C::MTP40C(Stream * str) : MTP40(str)
{
  _type = 2;
};


MTP40D::MTP40D(Stream * str) : MTP40(str)
{
  _type = 3;
};


// -- END OF FILE --

